toolkit/tools/imagegen/configuration/packagerepo.go

```diff
@@ -0,0 +1,180 @@
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+
+	// Now validate the resulting unmarshaled object
+	err = p.IsValid()
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+	return
+}
+
+// IsValid returns an error if the PackageRepo struct is not valid
+func (p *PackageRepo) IsValid() (err error) {
+	err = p.nameIsValid()
+	if err != nil {
+		return
+	}
+
+	err = p.repoUrlIsValid()
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// UpdatePackageRepo creates additional repo files specified by image configuration
+// and returns error if the operation fails
+func UpdatePackageRepo(installChroot *safechroot.Chroot, config SystemConfig) (err error) {
+	const (
+		repoFileDir   = "/etc/yum.repos.d/"
+		localRepoFile = "/etc/yum.repos.d/mariner-iso.repo"
+		squashErrors  = false
+	)
+
+	if exists, ferr := file.DirExists(repoFileDir); ferr != nil {
+		logger.Log.Errorf("Error accessing repo file directory.")
+		err = ferr
+		return
+	} else if !exists {
+		err = fmt.Errorf("%s: no such directory", repoFileDir)
+		return
+	}
+
+	// Remove mariner-iso.repo
+	os.Remove(localRepoFile)
+
+	// Loop through the PackageRepos field to determine if any customized package repos are specified.
+	// If specified, create new repo files for them
+	for _, packageRepo := range config.PackageRepos {
+		err = createCustomPackageRepo(installChroot, packageRepo, repoFileDir)
+		if err != nil {
+			return
+		}
+	}
+
+	// It is possible that network access may not be up at this point,
+	// so check network access
+	err = network.CheckNetworkAccess()
+	return
+}
+
+// nameIsValid returns an error if the package repo name is empty
+func (p *PackageRepo) nameIsValid() (err error) {
+	if strings.TrimSpace(p.Name) == "" {
+		return fmt.Errorf("invalid value for package repo name (%s), name cannot be empty", p.Name)
+	}
+	return
+}
+
+// repoUrlIsValid returns an error if the package url is invalid
+func (p *PackageRepo) repoUrlIsValid() (err error) {
+	if strings.TrimSpace(p.BaseUrl) == "" {
+		return fmt.Errorf("invalid value for package repo URL (%s), URL cannot be empty", p.BaseUrl)
+	}
+
+	_, err = url.ParseRequestURI(p.BaseUrl)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func writeAdditionalFieldstoRepoFile(fileName string, stringBuilder *strings.Builder) {
+	const (
+		gpgKey       = "gpgkey=file:///etc/pki/rpm-gpg/MICROSOFT-RPM-GPG-KEY file:///etc/pki/rpm-gpg/MICROSOFT-METADATA-GPG-KEY\n"
+		enable       = "enabled=1\n"
+		gpgCheck     = "gpgcheck=1\n"
+		repogpgCheck = "repo_gpgcheck=1\n"
+		skip         = "skip_if_unavailable=True\n"
+		sslVerify    = "sslverify=1\n"
+	)
+
+	additionalFields := gpgKey + enable + gpgCheck + repogpgCheck + skip + sslVerify
+	stringBuilder.WriteString(additionalFields)
+}
+
+func createCustomRepoFile(fileName string, packageRepo PackageRepo) (err error) {
+	var stringBuilder strings.Builder
+
+	err = file.Create(fileName, 0644)
```

Comment:
Looking through the tests reminded me of another thing. We should make sure functions don't leave any leftovers if they happen to hit an error. In this case, any error after a successful creation of the file should trigger a removal of that file. Caller's expectation from a failing function is that it didn't change the state of the system. Sometimes that cannot be guaranteed but in this case it shouldn't be that hard and if we fail to remove the file after it got successfully created, then it's possible the system is in a greater peril anyway.:)

Anyway, please take a look at how the `defer` call inside `imager.go::setupLoopDeviceDisk()` makes sure the functions cleans up after itself and let's do something similar here and inside `createCustomPackageRepo()` below.

Reply:
Good point. Have added implementation to clean up repo files in case there are any errors

Reply:
Did you add that clean-up? I'm not sure where it is.

Reply:
i added it in the createCustomPackageRepo() function below